docs(main): fix server.proxy doc example code

luckynumberke7in opened the pull request:
docs(main): fix server.proxy doc example code

wrap `path.replace` regex in a string to resemble working code

should resolve issues like #1014, #3475, #10168

<!-- Thank you for contributing! -->

### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [ ] New Feature
- [x] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

luckynumberke7in committed:
docs(main): fix server.proxy doc example code

docs(main): fix server.proxy doc example code

wrap `path.replace` regex in a string to resemble working code

should resolve issues like #1014, #3475, #10168

luckynumberke7in commented:
![vite working](https://user-images.githubusercontent.com/60681675/191134189-9c9a297b-a05f-428e-8e0f-86dadf8240eb.PNG)

This is the working file, it had the escape character in it, but when I saved the string, it automatically disappeared (and still works like this).

![vite broken](https://user-images.githubusercontent.com/60681675/191134292-51a479f6-5980-4f99-be87-76adb8b6082f.PNG)

This is the broken version that the docs suggest.

You can see the 404 failure and that the proxy is not working.

I have also set up a proxy using direct path and no `rewrite:` which worked, and lead me to believe it was an issue with the rewrite path logic.

luckynumberke7in commented:
Note: the working one works with and without `secure: false,` and the broken version does not work with or without `secure: false`

luckynumberke7in commented:
Also, I think it's failing for the save discrepancy I mentioned above (when I saved the escape characters vanish with prettier).

luckynumberke7in commented:
I should probably update the PR to be `rewrite: (path) => path.replace('/^/api/', ''),` and `rewrite: (path) => path.replace('/^/fallback/', ''),` huh?

This is my first public PR, sorry if I'm all over the place!

luckynumberke7in commented:
For future reference: How would I go about updating my PR? Do I have to refork and redo? or can I push another change to the current PR?

tony19 commented:
The reason your change worked in your own code is likely because it effectively prevented the `rewrite` from occurring, which means you don't need the `rewrite` in the first place. Just remove `rewrite` in your config.

> For future reference: How would I go about updating my PR? Do I have to refork and redo? or can I push another change to the current PR?

You can continue pushing changes to the PR. If the PR is not ready to be reviewed, mark it as a Draft. In my PRs, if I expect a high number of pushes (which causes a lot of notification noise for reviewers), I make my changes in a branch off my PR branch, and then merge back when ready for review.

luckynumberke7in commented on docs/config/server-options.md:
```diff
@@ -95,13 +95,13 @@ export default defineConfig({
         target: 'http://jsonplaceholder.typicode.com',
         changeOrigin: true,
-        rewrite: (path) => path.replace(/^\/api/, '')
+        rewrite: (path) => path.replace('/^\/api/', '')
```
Ahh, that did work! Sorry, everything I found online included the rewrite, but it just was not working.

I came to the docs and came to the same conclusion since it was in options and regex.

Maybe it could be made a little more clear that proxy matches following routes (i.e. '/api/products', /api/products/1', etc.) by default using:  
```proxy: {
      '/api': {
        target: 'http://localhost:5000',
        changeOrigin: true,
        secure: false,
      }```
      
      It was confusing for me at least.

sapphi-red replied:
Yeah, I think we could have examples.
```
// string shorthand: http://localhost:5173/foo -> http://localhost:4567/foo
// with options: http://localhost:5173/api/bar-> http://jsonplaceholder.typicode.com/bar
```

luckynumberke7in replied:
for readability, I was thinking something more along the lines of:
```js
export default defineConfig({
  server: {
    proxy: {
      // string shorthand, matches calls to '/foo/bar' by default
      '/foo': 'http://localhost:4567',
      // with options, returns http://jsonplaceholder.typicode.com/bar
      '/api': {
        target: 'http://jsonplaceholder.typicode.com',
        changeOrigin: true,
        rewrite: (path) => path.replace(/^\/api/, '')
      },```

luckynumberke7in replied:
the matching subsequent folders by default is what was confusing me. I thought I needed a rewrite to achieve this.

bluwy commented:
@luckynumberke7in do you want to implement https://github.com/vitejs/vite/pull/10169#discussion_r975381491 in this PR? Marking this one as draft for now, once you're ready you can mark this ready for review.

luckynumberke7in committed:
docs(main): add examples to server.proxy doc

docs(main): add examples to server.proxy doc

add examples to proxy for readability

should resolve issues like #1014, #3475, #10168

sapphi-red commented on docs/config/server-options.md:
```diff
@@ -89,9 +89,9 @@ In some cases, you might also want to configure the underlying dev server (e.g.
+      // string shorthand, matches calls to '/foo/bar' by default
       '/foo': 'http://localhost:4567',
-      // with options
+      // with options, returns http://jsonplaceholder.typicode.com/bar
```
> the matching subsequent folders by default is what was confusing me. I thought I needed a rewrite to achieve this.

If this is the reason, I think it's better to update the description (line 80) because these examples are a bit ambiguous.
